AKS Endpoint Caller (#1)
mspnp/aks-endpoint-caller: README.md

It is a very simple application that allows us to test communication and lack of it.

It is mainly useful to test networking policies. It was created to help us on AKS Regulated Cluster, but it can be used base on your needs.

## Application
It is a basic .Net 5.0 web application. It was created using Visual Studio 2019.

### Build docker images
Before deploying on Kubernetes we need to create the docker image. There is a DockerFile as part of the solution.

```bash
docker build -f ".\SimpleChainApi\Dockerfile" --force-rm -t aks-endpoint-caller --target final .
```

If you are using Azure Kubernetes Service you need to [push your image on a Azure Container Registry](https://docs.microsoft.com/en-us/azure/container-registry/container-registry-get-started-docker-cli)

### Kubertnes deploy

It is an example of a deployment on AKS.
 
```yml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: ingress-endpoint
spec:
  selector:
    matchLabels:
      app: ingress-endpoint
  replicas: 1
  template:
    metadata:
      labels:
        app: ingress-endpoint
    spec:
      containers:
      - name: ingress-endpoint

>         image: myacr.azurecr.io/aks-endpoint-caller

Reviewer: ```suggestion
        image: myacr.azurecr.io/aks-endpoint-caller:1.0
```

Author: done

        ports:

>         - containerPort: 80

Reviewer: Let's keep it exposed on 8080 (adding the `http://+:8080` environment variable)

Author: I changed to 8080, but I didn't added a environment variable.

Reviewer: Any reason not to?  Just going to assume that the container's default environment variable for the same will be fine?

Author: I'm just worry for the lack of time to test. Tomorrow I need to ping Rick to start something else. If I can't test the new configuration I will keep the one that I'm sure is working and do it in the future.

Author: done, tested

        env:
        - name: SELF_HOSTS_DEPENDENCIES
          value: "http://service-a,http://service-b,http://service-c"
        - name: EXTERNAL_DEPENDENCIES
          value: "https://whatismyip.io/"
        - name: "DEPTH"
          value: "2"
```

> We need to define some environment variables:
> 1. EXTERNAL_DEPENDENCIES, These are URLs with an unknown result
> 1. SELF_HOSTS_DEPENDENCIES, These are URLs where instances of this micro-service is deployed, so the result format is well known

Reviewer: ```suggestion
We need to define some environment variables:
1. `EXTERNAL_DEPENDENCIES` - These are URLs with an unknown result; as a comma-separated list
1. `SELF_HOSTS_DEPENDENCIES` - These are URLs where instances of this micro service is deployed, so the result format is well known; as a comma-separated list.
```

Author: done

> 1. DEPTH, This value is for doing the recursive calls. The default is 0, not dependency is call. If the value is bigger than 0 the dependencies are call, and the recursive dependencies are called using (DEPTH-1). This parameter avoid infinite calls. This parameter is needed to be set only in the entry point of micro services components.

Reviewer: ```suggestion
1. `DEPTH` - This value is for doing the recursive calls. The default is `0`, meaning no dependencies are called. If the value is larger than `0` the dependencies are call, and the recursive dependencies are called using (`DEPTH-1`). This parameter avoid infinite calls. This parameter is needed to be set only in the entry point of micro services components.
```

Author: done

### Example

> We can create the configuration which is useful for us. A possible setup is:

Reviewer: ```suggestion
We can create the configuration which is useful for the situation we want to test. An example setup is:
```

Author: done

> 1. Ingress Enpoint is the starting point, and calls: https://www.microsoft.com, https://serviceA, https://serviceB,  https://serviceC
> 1.  serviceA calls: https://www.microsoft.com, https://serviceA, https://serviceB,  https://serviceC
> 1.  serviceB calls: https://www.microsoft.com, https://serviceA, https://serviceB,  https://serviceC
> 1.  serviceC calls: https://www.microsoft.com, https://serviceA, https://serviceB,  https://serviceC   

Reviewer: ```suggestion
1. Ingress Endpoint is the starting point, and calls: https://www.microsoft.com, https://serviceA, https://serviceB,  https://serviceC
1.  serviceA calls: https://www.microsoft.com, https://serviceA, https://serviceB, https://serviceC
1.  serviceB calls: https://www.microsoft.com, https://serviceA, https://serviceB, https://serviceC
1.  serviceC calls: https://www.microsoft.com, https://serviceA, https://serviceB, https://serviceC   
```

Author: done
